Realm/RLMAccessor.mm

```diff
@@ -160,29 +162,64 @@ void setValue(__unsafe_unretained RLMObjectBase *const obj, ColKey key,
     }
 }
 
+// set getter/setter
+RLMSet *getSet(__unsafe_unretained RLMObjectBase *const obj, NSUInteger propIndex) {
+    RLMVerifyAttached(obj);
+    auto prop = obj->_info->rlmObjectSchema.properties[propIndex];
+    return [[RLMManagedSet alloc] initWithParent:obj property:prop];
+}
+
 // array getter/setter
 RLMArray *getArray(__unsafe_unretained RLMObjectBase *const obj, NSUInteger propIndex) {
     RLMVerifyAttached(obj);
     auto prop = obj->_info->rlmObjectSchema.properties[propIndex];
     return [[RLMManagedArray alloc] initWithParent:obj property:prop];
 }
 
-void setValue(__unsafe_unretained RLMObjectBase *const obj, ColKey key,
-              __unsafe_unretained id<NSFastEnumeration> const value) {
-    auto prop = obj->_info->propertyForTableColumn(key);
-    RLMValidateValueForProperty(value, obj->_info->rlmObjectSchema, prop, true);
+template <typename T>
+auto get_value(T t) {
+    if constexpr (std::is_pointer_v<T>)
+        return *t;
+    else
+        return t;
+}
 
-    realm::List list(obj->_realm->_realm, obj->_row, key);
-    RLMClassInfo *info = obj->_info;
-    if (list.get_type() == realm::PropertyType::Object) {
+template <typename Collection>
+void assign_value(Collection&& collection,
+                  RLMClassInfo *info,
+                  __unsafe_unretained RLMObjectBase *const obj,
+                  __unsafe_unretained RLMProperty *const prop,
+                  __unsafe_unretained id<NSFastEnumeration> const value) {
+    if (collection.get_type() == realm::PropertyType::Object) {
         info = &obj->_info->linkTargetType(prop.index);
     }
     RLMAccessorContext ctx(*info);
     RLMTranslateError([&] {
-        list.assign(ctx, value, realm::CreatePolicy::ForceCreate);
+        collection.assign(ctx, value, realm::CreatePolicy::ForceCreate);
     });
 }
 
+void setValue(__unsafe_unretained RLMObjectBase *const obj, ColKey key,
+              __unsafe_unretained id<NSFastEnumeration> const value) {
+    auto prop = obj->_info->propertyForTableColumn(key);
+    RLMValidateValueForProperty(value, obj->_info->rlmObjectSchema, prop, true);
+
+    if (prop.array) {
```

Comment:
The compiler will complain about incompatible operand types

Reply:
List and Set should probably just share this functionality in core, in which case you wouldn't need to template the method and could just use a base type

Reply:
You could make a function that returns `auto` for this:
```objc
auto CollectionFromObject(__unsafe_unretained RLMObjectBase *const obj, ColKey key) {
    auto prop = obj->_info->propertyForTableColumn(key);
    if (prop.array) {
        return realm::List(obj->_realm->_realm, obj->_row, key);
    } else {
        realm::object_store::Set(obj->_realm->_realm,
                                 obj->_row, key);
    }
}
```
but I think if this is the only place it would be used, then it's not worth it.